src/controller.js

```diff
@@ -431,22 +448,43 @@ class Controller {
       throw new Error('Object must total fewer than 5 TB'); 
     }
 
-    let headers = this.__generateMetadataHeaders(metadata, {
+    let headers = {
       'content-sha256': sha256,
       'content-length': size,
-    });
+    };
+
+    // Figure out content-encoding and sha256 values
+    if (contentEncoding && contentEncoding !== 'identity') {
+      if (!transferSha256) {
+        throw new Error('When using Content-Encoding, a transferSha256 value must be given');
```

Comment:
you can still give it a default value, if you wish... just do so before making the assert.... I don't like this double layer of if statements here... it's a bit hard to read... Could we not get the same logic by setting defaults first, and then doing asserts later?